osu.Game.Rulesets.Osu/Difficulty/Preprocessing/L2NormCorrection.cs

```diff
@@ -0,0 +1,210 @@
+                    distance: snap0distances,
+                    x_offset: new double[] { 0.25, 1, 0.7, 2, 2, 2 },
+                    y_offset: new double[] { -0.5, -2, -2.8, -4, -6, -6 },
+                    offset:   new double[] { 1, 1, 1, 1, 1, 1 },
+                    scale:    new double[] { 0.6, 1, 0.8, 0.3, 0.2, 0.2 }
+                ),
+                new L2NormCorrection(
+                    distance: snap0distances,
+                    x_offset: new double[] { 0, 0, -0.5, -2, -3, -3 },
+                    y_offset: new double[] { 0, 0, 0, 0, 0, 0 },
+                    offset:   new double[] { 1, 1, 1, 1, 1, 1 },
+                    scale:    new double[] { -0.7, -1, -0.9, -0.1, -0.1, -0.1 }
+                )
+            });
+
+        private static readonly double[] flow3distances = { 0, 1, 2, 3, 4 };
+        public static readonly MultiL2NormCorrection FLOW_3 = new MultiL2NormCorrection(
+            distance: flow3distances,
+            offset: new double[] { -4, -5.3, -5.2, -2.5, -2.5 },
+            scale: new double[] { 1, 1, 1, 1, 1 },
+            components: new[] {
+                new L2NormCorrection(
+                    distance: flow3distances,
+                    x_offset: new double[] { 0, 1.2, 2, 2, 2 },
+                    y_offset: new double[] { 0, 0, 0, 0, 0 },
+                    offset:   new double[] { 0, 0, 0, 0, 0 },
+                    scale:    new double[] { 1.5, 1, 0.4, 0, 0 }
+                ),
+                new L2NormCorrection(
+                    distance: flow3distances,
+                    x_offset: new double[] { 0, 0, 0, 0, 0 },
+                    y_offset: new double[] { 0, 0, 0, 0, 0 },
+                    offset:   new double[] { 0, 0, 0, 0, 0 },
+                    scale:    new double[] { 2, 1.5, 2.5, 3.5, 3.5 }
+                ),
+                new L2NormCorrection(
+                    distance: flow3distances,
+                    x_offset: new double[] { 0, 0.3, 0.6, 0.6, 0.6 },
+                    y_offset: new double[] { 0, 1, 2.4, 2.4, 2.4 },
+                    offset:   new double[] { 0, 0, 0, 0, 0 },
+                    scale:    new double[] { 0, 0.4, 0.4, 0, 0 }
+                ),
+                new L2NormCorrection(
+                    distance: flow3distances,
+                    x_offset: new double[] { 0, 0.3, 0.6, 0.6, 0.6 },
+                    y_offset: new double[] { 0, -1, -2.4, -2.4, -2.4 },
+                    offset:   new double[] { 0, 0, 0, 0, 0 },
+                    scale:    new double[] { 0, 0.4, 0.4, 0, 0 }
+                ),
+            });
+
+        private static readonly double[] snap3distances = { 1, 1.5, 2.5, 4, 6, 8 };
+        public static readonly MultiL2NormCorrection SNAP_3 = new MultiL2NormCorrection(
+            distance: snap3distances,
+            offset: new double[] { -2, -2, -3, -5.4, -4.9, -4.9 },
+            scale: new double[] { 1, 1, 1, 1, 1, 1 },
+            components: new[] {
```

Comment:
i feel like amount of components isnt immediately obvious here, might want to have some sort of param to show how many components should be there and add checks when creating the class to ensure its correct

Reply:
there can be any number of "components", they're just summed up. If you mean the length of each array, I think the InterpolateSorted function should check that, though I'm not sure